qutebrowser/completion/models/filepathcategory.py

```diff
@@ -0,0 +1,63 @@
+# vim: ft=python fileencoding=utf-8 sts=4 sw=4 et:
+
+# This file is part of qutebrowser.
+#
+# qutebrowser is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# qutebrowser is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Completion category for filesystem paths."""
+
+import glob
+import os
+from typing import Any
+import urllib.parse
+
+from PyQt5.QtCore import (QAbstractListModel, QModelIndex)
+
+
+class FilePathCategory(QAbstractListModel):
+    """Represent filesystem paths matching a pattern."""
+
+    def __init__(self, name: str):
+        super().__init__()
+        self._paths: list = []
+        self.name = name
+        self.columns_to_filter = [0]
+
+    def set_pattern(self, val: str) -> None:
+        """Compute list of suggested paths. (Called from `CompletionModel`.)
+
+        Args:
+            val: The user's partially typed URL/path.
+        """
+        if not val:
+            self._paths = [os.path.expanduser('~')]
```

Comment:
I like the idea of offering the home directory here! A couple of thoughts:

- Should we really expand it, or just display `~` initially?
- Are there more useful directories we could add? Perhaps the downloads directory?
- Or should we even offer a config option so the user could add their preferred directories here? That seems like something that could be pretty useful, e.g. for people doing web development.

Reply:
> Should we really expand it, or just display ~ initially?

I'd say just `~`, and do the expansion behind-the-scenes, to keep paths shorter.

> offer a config option so the user could add their preferred directories here

I like this idea. It might be worth considering `completion.file_path.exclude` too, in case the user has something like an `sshfs` that is slow to access (would that hold up the entire completion?)

Reply:
Actually I'd say we should do here whatever we do later. If we can always keep `~` un-expanded (which would be ideal IMHO), we should show `~` here. If that's infeasible, then we should expand here too.